Error while checking build/O2/fullCI_slc9 for 5f7e8a19759ac9e60d10bd53c358e03d1f05286e at 2025-12-15 13:37:
```
## sw/BUILD/O2-latest/log
/sw/SOURCES/O2/14924-slc9_x86-64/0/DataFormats/Reconstruction/include/ReconstructionDataFormats/Vertex.h:49:3: error: unknown type name 'INVALID_TRIGGER_ERROR_NO_HOST_CODE'
/sw/SOURCES/O2/14924-slc9_x86-64/0/DataFormats/Reconstruction/include/ReconstructionDataFormats/Vertex.h:49:10: error: constructor cannot have a return type
/sw/SOURCES/O2/14924-slc9_x86-64/0/DataFormats/Reconstruction/include/ReconstructionDataFormats/Vertex.h:145:82: error: no matching constructor for initialization of 'VertexBase'
/sw/SOURCES/O2/14924-slc9_x86-64/0/DataFormats/Reconstruction/include/ReconstructionDataFormats/Vertex.h(51): error: calling a __device__ function("o2::math_utils::detail::GPUPoint2D<float, (int)0> ::SetX(float)") from a __host__ function("VertexBase") is not allowed
/sw/SOURCES/O2/14924-slc9_x86-64/0/DataFormats/Reconstruction/include/ReconstructionDataFormats/Vertex.h(52): error: calling a __device__ function("_ZN2o210math_utils6detail10GPUPoint2DI1?L1?0EE4SetYE1?") from a __host__ function("VertexBase") is not allowed
/sw/SOURCES/O2/14924-slc9_x86-64/0/DataFormats/Reconstruction/include/ReconstructionDataFormats/Vertex.h(53): error: calling a __device__ function("_ZN2o210math_utils6detail10GPUPoint3DI1?L1?0EE4SetZE1?") from a __host__ function("VertexBase") is not allowed
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14924/5f7e8a19759ac9e60d10bd53c358e03d1f05286e/build_O2_fullCI_slc9/pretty.html).

--- a/DataFormats/Reconstruction/include/ReconstructionDataFormats/Vertex.h
+++ b/DataFormats/Reconstruction/include/ReconstructionDataFormats/Vertex.h
@@ -45,9 +45,17 @@ class VertexBase
   static constexpr int kNCov = 6;
   GPUhdDefault() VertexBase() = default;
   GPUhdDefault() ~VertexBase() = default;
-  GPUhd() VertexBase(const math_utils::Point3D<float>& pos, const std::array<float, kNCov>& cov) : mPos(pos), mCov(cov)
+  GPUhd() VertexBase(const float* pos, const float* cov)
   {
+    mPos = math_utils::Point3D<float>(pos[0], pos[1], pos[2]);
+    mCov[kCovXX] = cov[kCovXX];
+    mCov[kCovXY] = cov[kCovXY];
+    mCov[kCovXZ] = cov[kCovXZ];
+    mCov[kCovYY] = cov[kCovYY];
+    mCov[kCovYZ] = cov[kCovYZ];
+    mCov[kCovZZ] = cov[kCovZZ];
   }
+  GPUhd() VertexBase(const math_utils::Point3D<float>& pos, const std::array<float, kNCov>& cov) : mPos(pos), mCov(cov) {}
 
 #if !defined(GPUCA_NO_FMT) && !defined(GPUCA_GPUCODE_DEVICE)
   void print() const;
@@ -58,6 +66,7 @@ class VertexBase
   GPUhd() float getX() const { return mPos.X(); }
   GPUhd() float getY() const { return mPos.Y(); }
   GPUhd() float getZ() const { return mPos.Z(); }
+  GPUhd() float getR() const { return gpu::CAMath::Hypot(mPos.X(), mPos.Y()); }
   GPUd() float getSigmaX2() const { return mCov[kCovXX]; }
   GPUd() float getSigmaY2() const { return mCov[kCovYY]; }
   GPUd() float getSigmaZ2() const { return mCov[kCovZZ]; }
@@ -69,6 +78,7 @@ class VertexBase
   GPUd() float getSigmaZ() const { return gpu::CAMath::Sqrt(getSigmaZ2()); }
 
   GPUd() const std::array<float, kNCov>& getCov() const { return mCov; }
+  GPUd() float getCov(int e) const { return mCov[e]; }
 
   GPUd() math_utils::Point3D<float> getXYZ() const { return mPos; }
   GPUd() math_utils::Point3D<float>& getXYZ() { return mPos; }
@@ -105,6 +115,7 @@ class VertexBase
     setSigmaYZ(syz);
   }
   GPUd() void setCov(const std::array<float, kNCov>& cov) { mCov = cov; }
+  GPUd() void setCov(float c, int e) { mCov[e] = c; }
 
   bool operator==(const VertexBase& other) const;
   bool operator!=(const VertexBase& other) const { return !(*this == other); }
@@ -133,10 +144,8 @@ class Vertex : public VertexBase
 
   GPUhdDefault() Vertex() = default;
   GPUhdDefault() ~Vertex() = default;
-  GPUhd() Vertex(const math_utils::Point3D<float>& pos, const std::array<float, kNCov>& cov, ushort nCont, float chi2)
-    : VertexBase(pos, cov), mChi2(chi2), mNContributors(nCont)
-  {
-  }
+  GPUhd() Vertex(const float* pos, const float* cov, ushort nCont, float chi2) : VertexBase(pos, cov), mChi2(chi2), mNContributors(nCont) {}
+  GPUhd() Vertex(const math_utils::Point3D<float>& pos, const std::array<float, kNCov>& cov, ushort nCont, float chi2) : VertexBase(pos, cov), mChi2(chi2), mNContributors(nCont) {}
 
 #if !defined(GPUCA_NO_FMT) && !defined(GPUCA_GPUCODE_DEVICE)
   void print() const;

